### ISSUE TITLE
as a user, I expect some feedback when some sort of async process is running

### ISSUE THREAD

**Issue by hamilton** [iodide-project/iodide#1053]
This is based on a comment @mdboom made in a PR:

> As an aside, one thing we could easily do given this refactor is: post the input code to the console pane immediately and show a spinner until cell execution is done, and then show the output. Right now (and previous to this change), nothing goes into the history until execution is done.

I know we've talked about this before, but this feels like a pretty serious UX issue that we know has already caused genuine confusion, and the fix is quite straightforward.

**Comment by hamilton** [comment]
@mdboom I'm thinking there are two places where this bit of feedback can occur:

1.) bottom right corner (or something like the bottom right), make a small overlay that has a spinner saying "running <spinner />" or something like that. Keeping this in the editor frame means it can't get clobbered by threadblocking on browsers that have implemented async iframes (chrome, etc.)
2.) in the console itself, for non-blocking async calls that are preventing further cell evaluation (`fetch` wrapped in `ioide.evalQueue.await`, plugin loads, etc.)

I'm wondering which one is a better focus, or whether these are separate things entirely. in the `iframe` world we're still blocked by expensive computations (even if it doesn't block the editor frame), so it seems that just implementing this with (1) might be the right approach.

**Comment by mdboom** [comment]
I think (1) feels safer.  I think we already have the needed messages passing between the iframe and the editor frame to make this work.

FWIW, Jupyter puts this in the upper right, which seems as good a place as any:

![image](https://user-images.githubusercontent.com/38294/47453267-4f4a3980-d79a-11e8-9283-f0514cc89b47.png)

It's not a spinner -- it's just a spinner, just a circle that is hollow or filled based on whether the kernel is busy.

**Comment by hamilton** [comment]
Todos on this:

- implement a kernel status symbol in the toolbar. Values should be `KERNEL_LOADING`, `KERNEL_ERROR`, `KERNEL_IDLE`, `KERNEL_BUSY`. Happy to change the terminology if this doesn't seem right, but we'll need something. This'll go in the toolbar next to the user menu.
- default kernel state is `KERNEL_LOADING`, which gets set to `KERNEL_IDLE` once the eval frame has loaded.
- wire up so that a cell evaluation of any kind updates state to `KERNEL_BUSY`. Any sort of resolution (error or otherwise) should set to `KERNEL_IDLE`.
- in an async iframe world (currently chrome only, eventually all browsers) any evaluation blocking async process should just be `KERNEL_BUSY`.

**Comment by hamilton** [comment]
THis will be fixed with #1250 / #1276

**Comment by bcolloran** [comment]
those PRs have merged, closing

**Comment by PythonLinks** [comment]
And when the task completes, some notification please. 
Specifically on save, ti would be great to be notified that it was saved.

**Comment by bcolloran** [comment]
@PythonLinks can you clarify what you need in a new issue? Do you mean you want a notification that your notebook has been saved? Because notebooks are saved automatically to the Iodide server every few seconds. Or do you mean you want confirmation of saving a file using our saving API? or something else? Thanks!!

**Comment by PythonLinks** [comment]
Great question.

It is interesting.  The problem was with my mental model. 
I am an old guy.  My thinking is that disk space is expensive.  My default assumption is 
that it only saves when I tell it to.  Sure I read somewhere that it saves automatically, 
automatically.   But I never assimilated that information. 

Plus I do things differently on the Forest Wiki.   Here let me give a 
different opinion. 

Saving automatically is great, until you want to restore.  Then you have an 
infinite list of versions.  Which is the right version?   What a pain to figure out.
Imagine if git saved every few seconds....

On the forest wiki, I make a bunch of edits, when I think it is 
good, I do a save. 
For difficult code, I give each version a different name. 
Then when I want to restore, there are not that many versions 
to sort through
and they are often labelled.

So what is the lesson?  If it auto  saves some visual feedback would be useful.
For my forest wiki , I should make it ask if a user wants to leave an edited page without saving.
I hope that answers your question.
